### PR TITLE
Add multiple-document support.

### LINKED ISSUE #4: Parser should account for multiple documents and throw error when ID already registered.
Note that the following concerns https://github.com/dojo/parser/blob/master/src/parser.ts#L77-L79. Considering that the parser accounts for multiple documents simultaneously, the `idMap` should as well, especially since node IDs are only supposed to be unique on a per-document basis. As for IDs that have already been registered within a single document, I'm inclined to believe an error should be thrown. Otherwise users might get into trouble without an immediate indication of the problem. For example, consider the following scenario:

```html
<div id="root">
	<p id="child"></p>
</div>
```

```ts
const root = document.getElementById('root');
watch({
	root: root
});

const oldChild = document.getElementById('child');
const newChild = document.createElement('div');
newChild.id = 'child';

root.insertBefore(newChild, oldChild);
root.removeChild(oldChild);
```

Since the new node is added before the old node is removed, `instantiateParserObject` will try to add the new instance to the `idMap` before the old one has been removed. As a result, `byNode(newChild)` will work, but `byId(newChild.id)` will not.

### PR DESCRIPTION
**Pull request by mwistrand**
Resolves #4. Allows widgets to be registered to multiple documents, without the possibility of ID collisions.

### CODE REVIEW

File: src/parser.ts [hunk 1]
```diff
@@ -73,10 +73,17 @@ function instantiateParserObject(node: HTMLElement, reject: ParserRejector): Par
 			instance = new Ctor(node, options);
 			instance.node = node;
 			if (node.id) {
+				let idMap = idDocumentMap.get(node.ownerDocument);
+				if (!idMap) {
+					idDocumentMap.set(node.ownerDocument, (idMap = <{ [id: string]: ParserObject }>{}));
```
Comment (kitsonk) [inline src/parser.ts]: The linter doesn't like the cuddled type with empty object literal.

File: src/parser.ts [hunk 2]
```diff
@@ -73,10 +73,17 @@ function instantiateParserObject(node: HTMLElement, reject: ParserRejector): Par
...
+				}
 				instance.id = node.id;
 				if (!(instance.id in idMap)) {
 					idMap[instance.id] = instance;
-				} /* What to do if it is already there? Maybe have a map per document? */
+				}
+				else {
+					throw new Error('An instance has already been registered for ID "' + node.id + '".');
```
Comment (kitsonk) [inline src/parser.ts]: We should probably start using template strings.

File: src/parser.ts [hunk 3]
```diff
@@ -40,9 +40,9 @@ const registryDocumentMap = new WeakMap<Document, Registry<ParserObjectConstruct
 const nodeMap = new WeakMap<HTMLElement, ParserObject>();
 
 /**
- * A hash of the object ids and their objects
+ * A map of document objects to a hash of the object ids and their objects
  */
-const idMap: { [id: string]: ParserObject } = {};
+const idDocumentMap = new WeakMap<Document, { [id: string]: ParserObject }>();
```
Comment (kitsonk) [inline src/parser.ts]: Since we are using our idMap hash in several places now, maybe best to extract it to an interface.

File: src/parser.ts [hunk 4]
```diff
@@ -112,10 +125,12 @@ export function byNode<T extends ParserObject>(node: HTMLElement): T {
 /**
  * The public API for retrieving a parser object by ID
  * @param  {string}                 id The ID of the parser object to retrieve
+ * @param  {Document}               doc The optional Document. Defaults to `document`.
  * @return {T extends ParserObject}    The instance associated with the ID or `undefiend`
  */
-export function byId<T extends ParserObject>(id: string): T {
-	return <T> idMap[id];
+export function byId<T extends ParserObject>(id: string, doc: Document = document): T {
```
Comment (kitsonk) [inline src/parser.ts]: I know we are doing the right thing on the ordering of the "optional" argument last, but I wonder from a developer ergonomics if an overloaded function might not be in order here... (this is more a of discussion/question on my part versus a strong opinion):

```typescript
function isDocument(value: any): value is document {
    return value instanceof Document;
}

export function byId<T extends ParserObject(id: string): T;
export function byId<T extends ParserObject(doc: Document, id: string): T;
export function byId<T extends ParserObject(doc: Document | string, id?: string): T {
    const idMap = isDocument(doc) ? idDocumentMap.get(doc) : idDocumentMap.get(document);
    return idMap && <T> idMap[id || doc];
}
```

File: src/parser.ts (hunk 1 above)
Comment (mwistrand) [inline src/parser.ts]: Fixed with 1d0db5a.

File: src/parser.ts (hunk 2 above)
Comment (mwistrand) [inline src/parser.ts]: Thanks for catching that. Fixed with 1d0db5a.

File: src/parser.ts (hunk 3 above)
Comment (mwistrand) [inline src/parser.ts]: Done in 1d0db5a, but now I wonder if this should be a local interface only.

File: src/parser.ts (hunk 4 above)
Comment (mwistrand) [inline src/parser.ts]: If it doesn't bother you that the `id` argument would no longer be in the same position in each context, then I would be okay with such an update.

**Pull request merged**